pkg/virt-controller/watch/export/export.go

```diff
@@ -384,19 +393,143 @@ func (ctrl *VMExportController) handleIsSourcePvc(vmExport *exportv1.VirtualMach
+			return retry, err
+		}
+	}
+
+	return ctrl.updateVMExportPvcStatus(vmExport, pvcs, pod, service, inUse)
+
+}
+
+func (ctrl *VMExportController) handleIsVmSnapshot(vmExport *exportv1.VirtualMachineExport, service *corev1.Service) (time.Duration, error) {
+	var retry time.Duration
+	pvcs := make([]*corev1.PersistentVolumeClaim, 0)
+
+	vmSnapshot, exists, err := ctrl.getVmSnapshot(vmExport.Namespace, vmExport.Spec.Source.Name)
+	restoreableSnapshots := 0
+	if err != nil {
+		return 0, err
+	} else if exists {
+		pvcs, restoreableSnapshots, err = ctrl.handlePVCsForVirtualMachineSnapshot(vmExport, vmSnapshot)
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	pod, exists, err := ctrl.getExporterPod(vmExport)
+	if err != nil {
+		return 0, err
+	} else if !exists {
+		if len(pvcs) == restoreableSnapshots && restoreableSnapshots > 0 {
+			pod, err = ctrl.createExporterPod(vmExport, pvcs)
+			if err != nil {
 				return 0, err
-			} else {
+			} else if pod == nil {
 				return retry, nil
 			}
+
+			if err := ctrl.getOrCreateCertSecret(vmExport, pod); err != nil {
+				return 0, err
+			}
+		}
+	} else {
+		if err := ctrl.handlePodSucceededOrFailed(vmExport, pod); err != nil {
+			return retry, err
 		}
 	}
 
-	return ctrl.updateVMExportPvcStatus(vmExport, pvcs, pod, service, inUse)
+	return ctrl.updateVMExportPvcStatus(vmExport, pvcs, pod, service, false)
+}
+
+func (ctrl *VMExportController) handlePodSucceededOrFailed(vmExport *exportv1.VirtualMachineExport, pod *corev1.Pod) error {
+	if pod.Status.Phase == corev1.PodSucceeded || pod.Status.Phase == corev1.PodFailed {
+		// The server died or completed, delete the pod.
+		ctrl.Recorder.Eventf(vmExport, corev1.EventTypeWarning, exporterPodFailedOrCompletedEvent, "Exporter pod %s/%s is in phase %s", pod.Namespace, pod.Name, &pod.Status.Phase)
+		return ctrl.Client.CoreV1().Pods(vmExport.Namespace).Delete(context.Background(), pod.Name, metav1.DeleteOptions{})
+	}
+	return nil
+}
+
+func (ctrl *VMExportController) handlePVCsForVirtualMachineSnapshot(vmExport *exportv1.VirtualMachineExport, vmSnapshot *snapshotv1.VirtualMachineSnapshot) ([]*corev1.PersistentVolumeClaim, int, error) {
+	var content *snapshotv1.VirtualMachineSnapshotContent
+	var err error
+	pvcs := make([]*corev1.PersistentVolumeClaim, 0)
+	exists := false
+	totalVolumes := 0
+
+	if vmSnapshot.Status.VirtualMachineSnapshotContentName != nil && *vmSnapshot.Status.VirtualMachineSnapshotContentName != "" {
+		content, exists, err = ctrl.getVmSnapshotContent(vmSnapshot.Namespace, *vmSnapshot.Status.VirtualMachineSnapshotContentName)
+		if err != nil {
+			return pvcs, totalVolumes, err
+		} else if exists {
+			sourceVm := content.Spec.Source.VirtualMachine
+			totalVolumes = len(content.Status.VolumeSnapshotStatus)
+			for _, volumeBackup := range content.Spec.VolumeBackups {
+				if pvc, err := ctrl.getOrCreatePVCFromSnapshot(vmExport, &volumeBackup, sourceVm); err != nil {
+					return pvcs, totalVolumes, err
+				} else if pvc != nil {
```

Comment:
Similar to above, when/why would `getOrCreatePVCFromSnapshot` ever return `nil, nil`?

Reply:
I don't see a way it would do that, but I am just being explicit about the fact I am not appending a nil to the list. I can switch it to just `else`